middlewares/sticky_backend_handler.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/containous/traefik/types"
+	guuid "github.com/satori/go.uuid"
+	"github.com/serialx/hashring"
+	"github.com/vulcand/oxy/forward"
+)
+
+// StickinessConfig holds the parsed representation of a Stickiness object
+type StickinessConfig struct {
+	BackendName      string
+	CookieEncryptKey string
+	CookieName       string
+	Rules            []*template.Template
+	UseCookie        bool
+	UseIP            bool
+	UseRules         bool
+}
+
+// NewStickinessConfig creates a new StickinessConfig instance
+func NewStickinessConfig(s *types.Stickiness, backendName string, cookieName string) *StickinessConfig {
+	var useCookie, useIP, useRules bool
+	rules := make([]*template.Template, 0)
+	useDefault := true
+	if len(s.Rules) > 0 {
+		useRules = true
+		useDefault = false
+		for i, rule := range s.Rules {
+			t, err := template.New(guuid.NewV4().String()).Funcs(sprig.FuncMap()).Parse(rule)
+			if err != nil {
+				log.Errorf("Backend %s: failed to parse sticky rule with index %d, error: %s", backendName, i, err)
+				t = nil
+			}
+			rules = append(rules, t)
+		}
+	}
+	if s.IP {
+		useIP = true
+		useDefault = false
+	}
+	if s.Cookie || useDefault {
+		useCookie = true
+	}
+
+	return &StickinessConfig{
+		BackendName:      backendName,
+		CookieEncryptKey: s.CookieEncryptKey,
+		CookieName:       cookieName,
+		Rules:            rules,
+		UseCookie:        useCookie,
+		UseIP:            useIP,
+		UseRules:         useRules,
+	}
+}
+
+// StickyBackendHandler inspects criteria such as Client IP address and uses
+// consistent hashing to route a request to a specific backend server
+type StickyBackendHandler struct {
+	config *StickinessConfig
+	lb     healthcheck.LoadBalancer
+	next   http.Handler
+}
+
+// NewStickyBackendHandler creates a new StickyBackendHandler instance
+func NewStickyBackendHandler(lb healthcheck.LoadBalancer, next http.Handler, config *StickinessConfig) *StickyBackendHandler {
+	return &StickyBackendHandler{lb: lb, next: next, config: config}
+}
+
+// ServeHTTP inspects the request for a sticky criteria and inserts the
+// sticky header if a criteria is met.  It then invokes the next handler
+// in the middleware chain.
+func (h *StickyBackendHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+
+	var server string
+	var key string
+	requestCookie, requestCookieErr := r.Cookie(h.config.CookieName)
+	stripSetCookie := true
+
+	if h.config.UseRules {
+		for i, rule := range h.config.Rules {
+			if rule != nil {
+				var b bytes.Buffer
+				var w = bufio.NewWriter(&b)
+				err := rule.Execute(w, r)
+				if err != nil {
+					log.Errorf("Backend %s: failed to execute sticky rule with index %d, error: %s", h.config.BackendName, i, err)
+				}
+				w.Flush()
+				key = strings.TrimSpace(b.String())
+				if key != "" {
+					break
+				}
+			}
+		}
+	}
+
+	if key == "" && h.config.UseIP {
+		var clientIP string
+		if xForwardedFor := r.Header.Get(forward.XForwardedFor); xForwardedFor != "" {
```

Comment:
Any reason why you only use X-Forwarded-For and not X-Real-IP?  While the X-Forwarded-For is a more common variant, the X-Real-IP is also seeing substantial usage, depending on what LB/Proxy one uses. Also, there is a "wrinkle in the fabric" in a form of usage of PROXY protocol, which, when in use, provides the real information via a different element in the originating HTTP request instead of a header (see [Proxy Protocl](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt), which is commonly used as well, especially by AWS and GCP in their load balancers.  Traefik already supports this via use of `go-proxyproto` library.  (See `EntryPoint.ProxyProtocol` config section).  So this might require adding more use-case handling.  @ldez, @emilevauge, @timoreimann - any thoughts?

Reply:
Good point, `X-Real-IP` needs to be here too.  It sounds like the Proxy Protocol case is handled already - is that underlying connection object swapped out before hitting the Sticky middleware?